statsmodels/tsa/statespace/_tools.pyx.in

```diff
@@ -21,26 +21,104 @@ cimport scipy.linalg.cython_lapack as lapack
 
 cdef FORTRAN = 1
 
+ctypedef fused sm_type_t:
+    np.float64_t
+    np.float32_t
+    np.complex64_t
+    np.complex128_t
+
+
+# ------------------------------------------------------------------------
 # Array shape validation
-cdef validate_matrix_shape(str name, Py_ssize_t *shape, int nrows, int ncols, nobs=None):
-    if not shape[0] == nrows:
+
+cdef validate_matrix_shape(str name, Py_ssize_t *shape, int nrows,
+                           int ncols, nobs=None):
+    if shape[0] != nrows:
         raise ValueError('Invalid shape for %s matrix: requires %d rows,'
                          ' got %d' % (name, nrows, shape[0]))
-    if not shape[1] == ncols:
+    if shape[1] != ncols:
         raise ValueError('Invalid shape for %s matrix: requires %d columns,'
                          'got %d' % (name, shape[1], shape[1]))
     if nobs is not None and shape[2] not in [1, nobs]:
         raise ValueError('Invalid time-varying dimension for %s matrix:'
                          ' requires 1 or %d, got %d' % (name, nobs, shape[2]))
 
-cdef validate_vector_shape(str name, Py_ssize_t *shape, int nrows, nobs = None):
-    if not shape[0] == nrows:
+
+cdef validate_vector_shape(str name, Py_ssize_t *shape, int nrows, nobs=None):
+    if shape[0] != nrows:
         raise ValueError('Invalid shape for %s vector: requires %d rows,'
                          ' got %d' % (name, nrows, shape[0]))
-    if nobs is not None and not shape[1] in [1, nobs]:
+    if nobs is not None and shape[1] not in [1, nobs]:
         raise ValueError('Invalid time-varying dimension for %s vector:'
                          ' requires 1 or %d got %d' % (name, nobs, shape[1]))
 
+
+# ------------------------------------------------------------------------
+# Blas Wrapping
+
+cdef inline copy(int size, sm_type_t* src_arr, int incx,
+                 sm_type_t* target, int incy):
+    """
+    Parameters
+    ----------
+    size : int
+        number of elements in vectors `src_arr` and `target`
+    src_arr : sm_type_t*
+        Pointer to Array that we are copying _from_
+    incx : int
+        Specifies the increment for the elements of `src_arr`
+    target : sm_type_t*
+        Pointer to Array that we are copying _to_
+    incy : int
+        Specifies the increment for the elements of `target`
+
+    References
+    ----------
+    https://software.intel.com/en-us/mkl-developer-reference-c-cblas-copy
+    """
+    if sm_type_t is np.float32_t:
+        blas.scopy(&size, src_arr, &incx, target, &incy)
+    elif sm_type_t is np.float64_t:
+        blas.dcopy(&size, src_arr, &incx, target, &incy)
+    elif sm_type_t is np.complex64_t:
+        blas.ccopy(&size, src_arr, &incx, target, &incy)
+    elif sm_type_t is np.complex128_t:
+        blas.zcopy(&size, src_arr, &incx, target, &incy)
+
+
+cdef inline swap(int size, sm_type_t* src_arr, int incx,
```

Comment:
We can do this for other functions, particularly to avoid having the repeated  `# Note: zdot and cdot are broken, so have to use gemv for those`